sterm.1
.TH STERM 1 "12 November 2023" "6.0.3" "sterm Manual"
.SH NAME
sterm \- a minimal serial terminal that focus on being easy to use. A client simply works.

.SH SYNOPSIS
.B sterm
[\fB\-h\fR | \fB\-\-help\fR] 
[\fB\-n\fR | \fB\-\-noecho\fR]
[\fB\-\-escape \fIcharacter\fR]
[\fB\-\-binary\fR]
[\fB\-b \fIbaudrate\fR | \fB\-\-baudrate \fIbaudrate\fR]
[\fB\-f \fIformat\fR | \fB\-\-format \fIformat\fR]
.IR "device"
.br

.SH DESCRIPTION
sterm is a minimal serial terminal that focus on being easy to use and not sucking. - This client simply works.
.br
It has inline input and supports Unicode (utf-8).
Each key typed gets directly send to the remote device.
It writes whatever it receives to stdout so that also ANSI escape sequences work as expected.
.br
With the \fI--binary\fR option, the received data will be output byte wise as hexadecimal numbers.
Ideal for debugging.
.br
sterm has some internal commands (escape sequences), listed below.
To enter a command you first have to enter an escape character.
Entering the escape character (see \fI--escape\fR) starts a command input.
If this command is known it gets executed.
If the command is unknown the input gets rejected.
When entering the escape character a second time, the escape character gets send to the remove client.


.SH OPTIONS
.TP
.BR \-h ", " \-\-help
Display a help, version number and exit
.TP
.BR \-\-binary
Print hexadecimal values instead of Unicode characters. (Only applied on output, input will still be UTF-8)
.TP
.BR \-n ", " \-\-noecho
Use this flag to avoid echoing the keys the user pressed to stdout.
When this flag is set, each character gets send directly, except when entering the escape key.
Then you can enter a command (escape sequence) that sterm handles internal.
.TP
.BR \-\-escape " " \fIcharacter\fR
With this parameter, the escape character can be changed.
By default, it is the escape key (\fI"\\e"\fR).
.TP
.BR \-b " " \fIbaudrate\fB  ", " \-\-baudrate " " \fIbaudrate\fR
Define the baudrate used for receiving and transmitting.
.br
Default: 115200 baud.
.TP
.BR \-f " " \fIformat\fB  ", " \-\-format " " \fIformat\fR
Configuration-triple: xyz with x = byte length in bits {5,6,7,8}; y = parity {N,E,O}; z = stop bits {1,2}.
.br
Default: "8N1" - 8 data bits, no parity bits and 1 stop bit.
.TP
.BR \-w " " \fIlogfile\fB  ", " \-\-write " " \fIlogfile\fR
Write received data into a log file. ANSI Escape sequences will also be written into the file.
In binary mode (\fI--binary\fR) binary data gets written into the file.
Otherwise the data is UTF-8 encoded.
.TP
.BR \fIdevice\fR
.br
Serial I/O device to access

.SH EXIT STATUS
.TP
0
Everything OK
.TP
1
On error

.SH INTERNAL COMMANDS
.TP
.BR exit
Close sterm
.TP
.BR version
Show the version of sterm

.SH EXAMPLES
.nf
.B sterm /dev/ttyUART0

.fi
Connect to UART device with 115200 baud, 8 data bits, 1 stop bit and no parity bit (8N1)

.P
.B sterm --binary -b 9600 -f 7E2 /dev/ttyS0

.fi
Connect with 9600 baud, 7 data bits, 2 stop bits and even parity:

.SH AUTHOR
Written by Ralf Stemmer <ralf.stemmer@gmx.net>

.SH REPORTING BUGS
GitHub: <https://github.com/rstemmer/sterm/issues>

.SH COPYRIGHT
sterm  Copyright (C) 2013-2023  Ralf Stemmer <ralf.stemmer@gmx.net>
.br
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>.
.br
This program comes with ABSOLUTELY NO WARRANTY
.br
This is free software, and you are welcome to redistribute it
under certain conditions.

.SH SEE ALSO
Full documentation at: <https://github.com/rstemmer/sterm>



CHANGELOG
6.0.3 - 12.11.2023: Deprecated thread API updated (.isAlive() → .is_alive() since Python 3.8) by gizmo1904 (https://github.com/gizmo1904)
6.0.2 - 02.07.2023: --write function fixed by Sam Grove (https://github.com/sg-)
6.0.1 - 07.03.2020: Issue in setup.py fixed
6.0.0 - 07.03.2020: Code refactoring and python packaging 
                    --unbuffered got removed. All input are now unbuffered.
                    --noecho added to suppress echoing entered keys (as it was like in unbuffered mode)
5.2.0 - 18.07.2019: --write added to log received data into a log file
5.1.0 - 16.04.2019: Unified the escape-sequence start symbol to escape (\e)
5.0.1 - 06.04.2019: Fixed a bug that lead to mishandling "\n"-Only line breaks when in unbuffered mode
5.0.0 - 12.01.2019: --unbuffered and --escape added to allow unbuffered input data transmission



sterm/sterm.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

# STERM, a serial communication terminal with server capabilities        #
# Copyright (C) 2013-2019  Ralf Stemmer (ralf.stemmer@gmx.net)           #
#                                                                        #
# This program is free software: you can redistribute it and/or modify   #
# it under the terms of the GNU General Public License as published by   #
# the Free Software Foundation, either version 3 of the License, or      #
# (at your option) any later version.                                    #
#                                                                        #
# This program is distributed in the hope that it will be useful,        #
# but WITHOUT ANY WARRANTY; without even the implied warranty of         #
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the          #
# GNU General Public License for more details.                           #
#                                                                        #
# You should have received a copy of the GNU General Public License      #
# along with this program.  If not, see <http://www.gnu.org/licenses/>.  #

import sys
import os
import tty
import termios
import time
import argparse
from threading import Thread
from uart       import UART, UARTMode
from terminal   import Terminal



VERSION = "6.0.0-indev"


# This global variable is used to shutdown the thread used
# for the ReceiveData function.
ShutdownReceiver = False


cli = argparse.ArgumentParser(
    description="A minimal serial ternimal.",
    formatter_class=argparse.ArgumentDefaultsHelpFormatter
    )

cli.add_argument(      "--binary",      default=False,                action="store_true",
    help="Display raw data instead of UTF-8 encoded. (read only)")
cli.add_argument("-u", "--unbuffered",  default=False,                action="store_true",
    help="Direct character transmission. Does not buffer a whole line for the user input.")
cli.add_argument(      "--escape",      default="\033",     type=str, action="store",
    help="Change the default escape character (escape)")
cli.add_argument("-b", "--baudrate",    default=115200,     type=int, action="store",
    help="The baudrate used for the communication.")
cli.add_argument("-f", "--format",      default="8N1",      type=str, action="store",
    help="Configuration-triple: xyz with x=bytelength in bits {5,6,7,8}, y=parity {N,E,O}, z=stopbits {1,2}.")
cli.add_argument("-w", "--write",       metavar="logfile",  type=str, action="store",
    help="Write received data into a file")
cli.add_argument("device",                                  type=str, action="store",
    help="Path to the serial communication device.")

UNBUFFERED = False      # Enable the unbuffered mode
ESCAPECHAR = "\033"     # Escape character to start an escape command sequence


def ReceiveData(uart, term):
    """
    This function reads every 0.1 seconds all data from the serial input buffer.
    The read data then gets printed to the screen (stdout).
    After writing to the output buffer, the buffer gets flushed so that the data is visible to the user
    as soon as possible.
    The function is blocking and runs until the global variable ``ShutdownReveiver`` get set to ``False``
    or when reading from the input buffer raises an exception.

    The behavior of this function depends on the ``binary`` argument:

    In *binary mode* the received bytes get encoded as one byte hexadecimal numbers with ``0x`` prefix,
    space separated and a space at the end of the byte stream.
    So when receiving the two bytes ``23``, ``42`` the output is ``0x23 0x42 ``.

    In *UTF-8 mode* the received data gets interpreted as UTF-8 encoded Unicode string.
    When an UnicodeDecodeError-Exception occurs, the raw data gets printed between ``[]``.

    Is the ``logfile`` parameter a sting, then all received data gets written into that file.
    In *binary mode* the data gets stored binary, otherwise it gets stored UTF-8 encoded.
    Also ANSI-Escape-Sequences will be stored in the file.
    The file gets opened in *append mode*. Old data will not be overwritten.

    This function is intended to run in a separate thread.
    The following example shows how to handle this function.

    .. code-block::

        # Start receiver thread
        ReceiverThread = Thread(target=ReceiveData, args=(uart, ))
        ReceiverThread.start()

        # …

        # Shutdown receiver thread
        ShutdownReceiver = True
        if ReceiverThread.isAlive():
            ReceiverThread.join()


    Args:
        uart: Instance of the *pyserial* ``Serial`` class.

    Returns:
        *Nothing*
    """

    data = ""
    while not ShutdownReceiver:

        string = uart.Receive()
        term.Write(string)
        time.sleep(0.1)



def ReadCommand(term):
    """
    Returns:
        A command string
    """
    char    = ""
    command = ""

    while True:
        char = term.ReadCharacter()
        if char == "\r":
            break
        elif char == ESCAPECHAR:
            if len(command) == 0:
                command = ESCAPECHAR
            break
        else:
            term.Write(char)
            command += char

    term.Write("\n")
    return command

def HandleUnbufferedUserInput(uart, term):
    """
    This function handles the user input in *Unbuffered Mode*.
    It reads the input from *stdin* byte by byte and sends it directly UTF-8 encoded to the UART device.

    When the escape character gets entered, the function starts to record a command instead of sending
    the data to the UART device.
    The escape character is defined in the global variable ``ESCAPECHAR`` that can be set via the
    command line parameter ``--escape``.
    The default character is *escape* (``\e``).
    Valid escape commands are ``exit`` to leave this function and exit ``sterm``,
    or ``version`` to print the version number of ``sterm`` to *stdout*.
    Enter the escape character twice send one escape character to the UART device.

    When the TTY is set to unbuffered mode, it expects the sequence ``\r\n`` for line breaks.
    This function takes care the ``\r\n`` sequences and ``\n``-only line breaks are handled correctly.

    The function expects UTF-8 encoded input.

    Args:
        uart: Instance of the *pyserial* ``Serial`` class.

    Returns:
        *Nothing*
    """
    char = ""

    while True:
        char = term.ReadCharacter()

        # Handle escape sequences
        if char == ESCAPECHAR:
            command = ReadCommand(term)

            if command == ESCAPECHAR:
                uart.write(ESCAPECHAR.encode("utf-8"))

            if command == "exit":
                break

            elif command == "version":
                print("Version: " + VERSION, end="\r\n")

        # Send character to UART-Device
        else:
            # In this mode, \n needs to be added manually to get a new line
            if char == "\r":
                char += "\n"
            uart.Transmit(char)

    return



def HandleBufferedUserInput(uart, term):
    r"""
    This function handles the user input in *Bufferd Mode*.
    It reads a complete line until the user hits the enter-key into a buffer.
    Then the whole buffer gets send UTF-8 encoded to the UART device.

    This function provides to escape commands.
    The escape character is defined in the global variable ``ESCAPECHAR`` that can be set via the
    command line parameter ``--escape``.
    The default character is *escape* (``\e``).
    When the line is ``"\033exit"`` this function gets left and ``sterm`` will shut down.
    On ``"\033version"`` the version number gets printed to *stdout*.

    The function expects UTF-8 encoded input.

    Args:
        uart: Instance of the *pyserial* ``Serial`` class.

    Returns:
        *Nothing*
    """
    command = ""

    while True:
        command = term.ReadLine()

        if command == ESCAPECHAR + "exit":
            break

        elif command == ESCAPECHAR + "version":
            print("Version: " + VERSION)

        else:
            command += "\n"
            uart.Transmit(command)
    return



def main():
    print("\n\033[1;31m --[ \033[1;34msterm \033[1;31m//\033[1;34m " + VERSION + "\033[1;31m ]-- \033[0m\n")

    # Handle command line arguments
    args       = cli.parse_args()
    DEVICE     = args.device
    BAUDRATE   = args.baudrate
    FORMAT     = args.format
    UNBUFFERED = args.unbuffered
    ESCAPECHAR = args.escape

    if args.binary:
        UARTMODE = UARTMode.BINARY
    else:
        UARTMODE = UARTMode.TEXT

    # Open remote terminal device
    uart = UART(args.device, args.baudrate, args.format, uartmode=UARTMODE, logpath=args.write)
    term = Terminal(buffered= not args.unbuffered, escape=args.escape)

    # Start receiver thread
    ReceiverThread = Thread(target=ReceiveData, args=(uart, term))
    ReceiverThread.start()

    # this is the main loop of this software
    try:
        if UNBUFFERED:
            HandleUnbufferedUserInput(uart, term);
        else:
            HandleBufferedUserInput(uart, term);
    except:
        # catch all to be able to clean up
        pass

    # Shutdown receiver thread
    global ShutdownReceiver
    ShutdownReceiver = True
    if ReceiverThread.isAlive():
        ReceiverThread.join()

    # Clean up everything
    uart.Disconnect()


if __name__ == '__main__':
    main()

# vim: tabstop=4 expandtab shiftwidth=4 softtabstop=4

